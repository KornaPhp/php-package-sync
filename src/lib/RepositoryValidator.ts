import { existsSync, mkdirSync } from 'fs';
import { app } from '../Application';
import PullPackageCommand from '../commands/PullPackage';
import PullTemplateCommand from '../commands/PullTemplate';
import { RepositoryKind } from './Repository';

export class RepositoryValidator {
    static ensureExists(name: string, kind: RepositoryKind) {
        if (kind === RepositoryKind.PACKAGE) {
            this.ensurePathExists(app.config.packagesPath);
            PullPackageCommand.handle({ name });
            return true;
        }

        if (kind === RepositoryKind.SKELETON) {
            this.ensurePathExists(app.config.templatesPath);
            PullTemplateCommand.handle({ name });
            return true;
        }

        return false;
    }

    static ensurePackageExists(name: string) {
        return this.ensureExists(name, RepositoryKind.PACKAGE);
    }

    static ensureTemplateExists(name: string) {
        return this.ensureExists(name, RepositoryKind.SKELETON);
    }

    protected static ensurePathExists(path: string) {
        if (path.length && !existsSync(path)) {
            mkdirSync(path, { recursive: true });
        }
    }
}
