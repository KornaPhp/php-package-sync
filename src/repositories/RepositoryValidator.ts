/* eslint-disable no-unused-vars */

import { existsSync, mkdirSync } from 'fs';
import PullPackageCommand from '../commands/PullPackage';
import PullTemplateCommand from '../commands/PullTemplate';
import { RepositoryKind } from './Repository';

//  (app.config.paths.packages,  app.config.paths.templates)
export class RepositoryValidator {
    constructor(public packagesPath: string, public templatesPath: string) {
        //
    }

    ensureExists(name: string, kind: RepositoryKind) {
        if (kind === RepositoryKind.PACKAGE) {
            this.ensurePathExists(this.packagesPath);
            PullPackageCommand.handle({ name });
            return true;
        }

        if (kind === RepositoryKind.SKELETON) {
            this.ensurePathExists(this.templatesPath);
            PullTemplateCommand.handle({ name });
            return true;
        }

        return false;
    }

    ensurePackageExists(name: string) {
        return this.ensureExists(name, RepositoryKind.PACKAGE);
    }

    ensureTemplateExists(name: string) {
        return this.ensureExists(name, RepositoryKind.SKELETON);
    }

    protected ensurePathExists(path: string) {
        if (path.length && !existsSync(path)) {
            mkdirSync(path, { recursive: true });
        }
    }
}
