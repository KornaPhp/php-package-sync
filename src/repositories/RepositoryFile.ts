/* eslint-disable no-unused-vars */

import { ComparisonScoreRequirements } from '../types/ComparisonScoreRequirements';
import { File } from '../lib/File';

export class RepositoryFile extends File {
    public shouldIgnore = false;
    public shouldCompare = false;
    public requiredScores: ComparisonScoreRequirements = { similar: 0, size: 0 };

    /**
     *
     * @param repository Repository class
     * @param fn string
     * @param contents string | null
     */
    constructor(public repository: any, protected fn: string, contents: string | null = null) {
        super(fn, contents);
    }

    get relativeName() {
        return this.fn.replace(this.repository.path + '/', '');
    }

    public withOptions(ignore: boolean, compare: boolean, requiredScores: ComparisonScoreRequirements) {
        this.shouldIgnore = ignore;
        this.shouldCompare = compare;
        this.requiredScores = requiredScores;
        //
        return this;
    }

    public processTemplate(): string {
        return super.processTemplate(this.repository.name);
    }
}
