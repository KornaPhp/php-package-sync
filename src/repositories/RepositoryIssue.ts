/* eslint-disable no-unused-vars */

import { ComparisonKind } from '../types/FileComparisonResult';
import { Repository } from './Repository';
import { RepositoryFile } from './RepositoryFile';
//import Fixer from './fixers/Fixer';
import { classOf } from '../lib/helpers';

export class RepositoryIssue {
    protected _fixers: any[] = [];
    public resolvedByFixer = 'none';
    public resolvedNotes: string[] = [];
    public pending = false;

    constructor(
        public result: any,
        public name: string,
        public srcFile: RepositoryFile | null,
        public destFile: RepositoryFile | null,
        public skeleton: Repository,
        public repository: Repository,
        public resolved: boolean = false,
        public note: string | null = null,
        public context: any | null = null,
    ) {
        //
    }

    get fixers() {
        const same = (a, b, method) => (a[method]() && b[method]()) || (!a[method]() && !b[method]());
        const isTrue = (a, method) => (a[method]() ? 1 : -1);
        const methodCompare = (a, b, method) => (same(a, b, method) ? 0 : isTrue(a, method));

        return this._fixers.sort((a, b) => methodCompare(a, b, 'runsFixers'))
            .sort((a, b) => methodCompare(a, b, 'isRisky'));
    }

    get kind(): ComparisonKind {
        return this.result.kind;
    }

    get score(): number | string {
        if (this.result.score === 0) {
            return '-';
        }

        if (typeof this.result.score === 'number') {
            return this.result.score.toFixed(3);
        }

        return this.result.score;
    }

    get sourcefile(): RepositoryFile {
        return <RepositoryFile>this.srcFile;
    }

    get targetfile(): RepositoryFile {
        return <RepositoryFile>this.destFile;
    }

    public addFixer(fixer: any) {
        if (this.fixers.find(f => f.getName() === fixer.getName()) === undefined) {
            this.fixers.push(fixer);
        }

        return this;
    }

    public is(kind: ComparisonKind): boolean {
        return this.kind === kind;
    }

    public resolve(resolvedByFixer: string | any) {
        if (typeof resolvedByFixer !== 'string') {
            resolvedByFixer = classOf(resolvedByFixer)
                .prettyName();
        }

        this.resolvedByFixer = <string>resolvedByFixer;
        this.resolved = true;

        return this;
    }

    public addResolvedNote(note: string, prepend = false) {
        if (prepend) {
            this.resolvedNotes.unshift(note);
            return this;
        }

        this.resolvedNotes.push(note);

        return this;
    }

    public disableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.disable();
            }
        });

        return this;
    }

    public enableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.enable();
            }
        });

        return this;
    }
}
