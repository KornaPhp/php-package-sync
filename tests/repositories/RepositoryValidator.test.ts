/* eslint-disable no-undef */
import { RepositoryValidator } from '../../src/repositories/RepositoryValidator';
import FakePullCommand from '../helpers/FakePullCommand';

beforeEach(() => {
    FakePullCommand.handled = [];
});

it('ensures a package has been cloned or updated', () => {
    const dataPath = __dirname + '/../data';
    const validator = new RepositoryValidator(dataPath, dataPath, FakePullCommand, FakePullCommand);

    validator.ensurePackageExists('my-test-package');

    expect(FakePullCommand.handled)
        .toMatchSnapshot();
});

it('ensures a template has been cloned or updated', () => {
    const dataPath = __dirname + '/../data';
    const validator = new RepositoryValidator(dataPath, dataPath, FakePullCommand, FakePullCommand);

    validator.ensureTemplateExists('my-test-template');

    expect(FakePullCommand.handled)
        .toMatchSnapshot();
});
