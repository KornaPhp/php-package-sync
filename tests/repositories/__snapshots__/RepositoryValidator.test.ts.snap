// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensures a package has been cloned or updated 1`] = `
[
  {
    "name": "my-test-package",
  },
]
`;

exports[`ensures a template has been cloned or updated 1`] = `
[
  {
    "name": "my-test-template",
  },
]
`;
